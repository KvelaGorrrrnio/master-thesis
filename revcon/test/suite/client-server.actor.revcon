@actor

global int workable_data

actor main.init() {
  spawn server server
  send server.setup()
}
actor main.serverup() {
  send server.connect()
}
actor main.serverdown() {
  despawn server server
}
actor main.connected(actor inWorker) {
  local actor worker
  swap inWorker worker
  send worker.work(workable_data)
}
actor main.disconnected() {
  send server.shutdown()
}
actor main.worked(int inData) {
  swap workable_data inData
  if workable_data > 101
  then 
    send server.disconnect(worker)
    delocal actor worker
  else send worker.work(workable_data)
  fi workable_data > 101
}

actor server.setup() from main {
  local list actor workers
  local actor worker
  local int i
  from i == 0
  do
    spawn worker worker
    push worker workers
    i += 1
  until i == 4
  delocal int i 4
  delocal actor worker

  send main.serverup()
}
actor server.shutdown() from main {
  local actor worker
  local int i
  from i == 0
  do
    pop worker workers
    despawn worker worker
    i += 1
  until i == 4
  delocal int i 4
  delocal actor worker
  delocal list actor workers

  send main.serverdown()
}
actor server.connect() from client {
  local actor worker
  pop worker workers
  send client.connected(worker)
  delocal actor worker
}
actor server.disconnect(actor worker) from client {
  push worker workers
  send client.disconnected()
}

actor worker.work(int data) from client {
  data += 10
  send client.worked(data)
}
