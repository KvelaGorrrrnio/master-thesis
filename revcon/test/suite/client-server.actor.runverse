test/suite/client-server.actor.revcon:
[store]
  !mailbox: [>MessageVal (-1) 0 "init" []>]
  !message: Ø
  !messages: []
  main: Ø

[stack]
[spawn main main | from UnOpEmpty (!mailbox) do 
[pop !message !messages | undeliver "!message" | undequeue !message !mailbox] <|> []
 loop 
[] <|> []
 until UnOpEmpty (!messages) | unsend Left "main".init() | despawn main main] <|> []


[procedures]
main.connected(inWorker)
main.disconnected()
main.init()
main.serverdown()
main.serverup()
main.worked(inData)
server.connect()
server.disconnect(worker)
server.setup()
server.shutdown()
worker.work(data)

[globals]
workable_data: 0
