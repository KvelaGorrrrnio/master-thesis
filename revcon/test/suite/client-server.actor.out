test/suite/client-server.actor.revcon:
[store]
  !mailbox: [>>]
  !message: Ø
  !messages: [MessageVal 1 0 "serverdown" [], MessageVal 0 1 "shutdown" [], MessageVal 1 0 "disconnected" [], MessageVal 0 1 "disconnect" [Ø], MessageVal 5 0 "worked" [0], MessageVal 0 5 "work" [0], MessageVal 5 0 "worked" [0], MessageVal 0 5 "work" [0], MessageVal 5 0 "worked" [0], MessageVal 0 5 "work" [0], MessageVal 5 0 "worked" [0], MessageVal 0 5 "work" [0], MessageVal 5 0 "worked" [0], MessageVal 0 5 "work" [0], MessageVal 5 0 "worked" [0], MessageVal 0 5 "work" [0], MessageVal 5 0 "worked" [0], MessageVal 0 5 "work" [0], MessageVal 5 0 "worked" [0], MessageVal 0 5 "work" [0], MessageVal 5 0 "worked" [0], MessageVal 0 5 "work" [0], MessageVal 5 0 "worked" [0], MessageVal 0 5 "work" [0], MessageVal 5 0 "worked" [0], MessageVal 0 5 "work" [0], MessageVal 1 0 "connected" [Ø], MessageVal 0 1 "connect" [], MessageVal 1 0 "serverup" [], MessageVal 0 1 "setup" [], MessageVal (-1) 0 "init" []]
  main: Ø

[stack]
[spawn main main | send Left "main".init() | from UnOpEmpty (!messages) do 
[dequeue !message !mailbox | deliver "!message" | push !message !messages] <|> []
 loop 
[] <|> []
 until UnOpEmpty (!mailbox) | despawn main main] <|> []


[procedures]
main.connected(inWorker)
main.disconnected()
main.init()
main.serverdown()
main.serverup()
main.worked(inData)
server.connect()
server.disconnect(worker)
server.setup()
server.shutdown()
worker.work(data)

[globals]
workable_data: 110
